compiler/rustc_middle/src/ty/context.rs

```diff
@@ -834,6 +834,7 @@ pub struct CanonicalUserTypeAnnotation<'tcx> {
     pub user_ty: CanonicalUserType<'tcx>,
     pub span: Span,
     pub inferred_ty: Ty<'tcx>,
+    pub variance: Variance,
```

Comment:
It's rather unclear to me what variance means here. I confess to have often "overloaded" it as a way to carry the expected relation between two types, but it seems necessary to comment the ordering here.

If you have "covariant", does that mean that the "inferred type" needs to be a subtype of the "user type"? etc